### PR TITLE
Full Foreign Key Referential Action Support & Self-Referential Support

### PR DESCRIPTION
**Pull request by Hydrocharged**
As stated in the title, this adds support for all of the referential foreign key actions, in addition to support for cyclical and self-referential foreign keys.

Tests are still failing as I haven't updated the existing tests enforcing incorrect behavior.

### CODE REVIEW

File: sql/analyzer/apply_foreign_keys.go [hunk 1]
```diff
@@ -160,57 +165,249 @@ func applyForeignKeysToNodes(ctx *sql.Context, a *Analyzer, n sql.Node, cache *f
...
+func getForeignKeyEditor(ctx *sql.Context, a *Analyzer, tbl sql.ForeignKeyTable, cache *foreignKeyTableCache, fkChain foreignKeyChain) (*plan.ForeignKeyEditor, error) {
+	fkEditor, err := getForeignKeyReferences(ctx, a, tbl, cache, fkChain)
+	if err != nil {
+		return nil, err
+	}
+	fkEditor2, err := getForeignKeyRefActions(ctx, a, tbl, cache, fkChain)
+	if err != nil {
+		return nil, err
+	}
+	// Due to tables not containing their database, we have to grab the table's database from the foreign key definition.
```
Comment (andrew-wm-arthur) [inline sql/analyzer/apply_foreign_keys.go:178]: I don't really follow what's happening here. Could we just add a `Database()` method to `sql.ForeignKeyTable`?

File: sql/analyzer/apply_foreign_keys.go [hunk 2]
```diff
@@ -160,57 +165,249 @@ func applyForeignKeysToNodes(ctx *sql.Context, a *Analyzer, n sql.Node, cache *f
...
+func getForeignKeyRefActions(ctx *sql.Context, a *Analyzer, tbl sql.ForeignKeyTable, cache *foreignKeyTableCache, fkChain foreignKeyChain) (*plan.ForeignKeyEditor, error) {
+	var editor sql.ForeignKeyUpdater
+	referencedFks, err := tbl.GetReferencedForeignKeys(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// MySQL has a CASCADE limit of a depth of 15
+	//TODO: figure out if MySQL errors once 15 has been hit, or if it just stops cascading
+	if len(referencedFks) == 0 || fkChain.Count() >= 15 {
+		return nil, nil
```
Comment (andrew-wm-arthur) [inline sql/analyzer/apply_foreign_keys.go:282]: We should error here for now, just to be safe

File: sql/analyzer/apply_foreign_keys.go [hunk 3]
```diff
@@ -160,57 +165,249 @@ func applyForeignKeysToNodes(ctx *sql.Context, a *Analyzer, n sql.Node, cache *f
...
+	}
+	// We can ignore foreign keys that have been previously used as we know that the data we're referencing is being inserted
+	for i := len(declaredFks) - 1; i >= 0; i-- {
+		if fkChain.HasForeignKey(declaredFks[i].Name) {
+			newDeclaredFks := make([]sql.ForeignKeyConstraint, 0, len(declaredFks)-1)
+			newDeclaredFks = append(newDeclaredFks, declaredFks[:i]...)
+			newDeclaredFks = append(newDeclaredFks, declaredFks[i+1:]...)
+			declaredFks = newDeclaredFks
+		}
+	}
```
Comment (andrew-wm-arthur) [inline sql/analyzer/apply_foreign_keys.go:208]:
```suggestion
	newDeclaredFks := make([]sql.ForeignKeyConstraint, 0, len(declaredFks))
	for _, fk := range declaredFks {
		if !fkChain.HasForeignKey(fk.Name) {
			newDeclaredFks = append(newDeclaredFks, fk)
		}
	}
	declaredFks = newDeclaredFks
```

File: sql/analyzer/apply_foreign_keys.go [hunk 4]
```diff
@@ -160,57 +165,249 @@ func applyForeignKeysToNodes(ctx *sql.Context, a *Analyzer, n sql.Node, cache *f
...
+}
+
+// getForeignKeyReferences returns an editor containing only the references for the given table.
+func getForeignKeyReferences(ctx *sql.Context, a *Analyzer, tbl sql.ForeignKeyTable, cache *foreignKeyTableCache, fkChain foreignKeyChain) (*plan.ForeignKeyEditor, error) {
+	var editor sql.ForeignKeyUpdater
+	declaredFks, err := tbl.GetDeclaredForeignKeys(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// We can ignore foreign keys that have been previously used as we know that the data we're referencing is being inserted
```
Comment (andrew-wm-arthur) [inline sql/analyzer/apply_foreign_keys.go:200]: If we've already added a foreign key to the `fkChain`, then we don't need to validate that its parent table contains the rows we're inserting?

File: sql/analyzer/apply_foreign_keys.go [hunk 5]
```diff
@@ -220,10 +417,82 @@ func (cache *foreignKeyTableCache) Get(ctx *sql.Context, a *Analyzer, dbName str
...
+	editor := foreignKeyTableUpdater{
+		tbl:     fkTbl,
+		updater: fkTbl.GetForeignKeyUpdater(ctx),
 	}
 	cache.cache[fkTableName] = editor
-	return editor.tbl, editor.editor, nil
+	return editor.tbl, editor.updater, nil
+}
+
+// foreignKeyChain holds all previously-referenced foreign keys and modified tables in the chain. Also keeps track of
```
Comment (andrew-wm-arthur) [inline sql/analyzer/apply_foreign_keys.go:428]: "previously-referenced" meaning previously used?

File: sql/plan/foreign_key_editor.go [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+func (fkEditor *ForeignKeyEditor) Update(ctx *sql.Context, old sql.Row, new sql.Row) error {
+	for _, reference := range fkEditor.References {
+		if err := reference.CheckReference(ctx, new); err != nil {
+			return err
+		}
+	}
+	for _, refActionData := range fkEditor.RefActions {
+		switch refActionData.ForeignKey.OnUpdate {
+		default: // RESTRICT and friends
+			err := func() error {
```
Comment (andrew-wm-arthur) [inline sql/plan/foreign_key_editor.go:71]: These closures here and in `Delete` would be better as standalone functions

**Review by andrew-wm-arthur — APPROVED** [review]
Structurally this looks really good. Nice work.
My comments are all around readability/simplicity, as Foreign Key logic is inherently complex, especially with all of the MySQL/InnoDB quirks. 
In my opinion, naming things with "parent"/"child" is easier to understand than "referenced"/"declared", even if the latter is closer to the language of SQL.

File: sql/analyzer/apply_foreign_keys.go (hunk 2 above)
Comment (Hydrocharged) [reply sql/analyzer/apply_foreign_keys.go:282]: To clarify the `TODO`, it's not an error here during analysis, but rather if a `CASCADE` actually hits a depth of 15 at runtime. So if it does error at 15, then I'd replace the `ForeignKeyEditor` node with a `ForeignKeyDepthError` node that just returns an error.

File: sql/analyzer/apply_foreign_keys.go (hunk 4 above)
Comment (Hydrocharged) [reply sql/analyzer/apply_foreign_keys.go:200]: Correct, as this child table would only be called as a result of a cascading referential action, so we can guarantee it's in the parent table.

File: sql/analyzer/apply_foreign_keys.go (hunk 3 above)
Comment (Hydrocharged) [reply sql/analyzer/apply_foreign_keys.go:208]: This is why PR reviews are good, as that is some horrendous code I wrote 🤣

**Comment by Hydrocharged** [comment]
Changes were manually added into the base PR.

**Pull request closed without merging**